infoschema/slow_log.go

```diff
@@ -0,0 +1,195 @@
+		}
+		rows = append(rows, record)
+	}
+	return rows, nil
+}
+
+// TODO: Support parse multiple log-files.
+// parseSlowLogFile uses to parse slow log file.
+func parseSlowLogFile(filePath string) ([]map[string]types.Datum, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer func() {
+		if err = file.Close(); err != nil {
+			log.Error(err)
+		}
+	}()
+
+	return parseSlowLog(bufio.NewScanner(file))
+}
+
+// TODO: optimize for parse huge log-file.
+func parseSlowLog(scanner *bufio.Scanner) ([]map[string]types.Datum, error) {
+	rows := make([]map[string]types.Datum, 0)
+	rowMap := make(map[string]types.Datum, len(slowQueryCols))
+	startFlag := false
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		// Check slow log entry start flag.
+		if !startFlag && strings.HasPrefix(line, variable.SlowLogStartPrefixStr) {
+			value, err := parseSlowLogField(variable.SlowLogTimeStr, line[len(variable.SlowLogStartPrefixStr):])
+			if err != nil {
+				log.Errorf("parse slow log error: %v", err)
+				continue
+			}
+			rowMap[variable.SlowLogTimeStr] = *value
+			startFlag = true
+			continue
+		}
+
+		if startFlag {
+			// Parse slow log field.
+			if strings.Contains(line, variable.SlowLogPrefixStr) {
+				line = line[len(variable.SlowLogPrefixStr):]
+				fieldValues := strings.Split(line, " ")
+				for i := 0; i < len(fieldValues)-1; i += 2 {
+					field := fieldValues[i]
+					if strings.HasSuffix(field, ":") {
+						field = field[:len(field)-1]
+					}
+					value, err := parseSlowLogField(field, fieldValues[i+1])
+					if err != nil {
+						log.Errorf("parse slow log error: %v", err)
+						continue
+					}
+					rowMap[field] = *value
+
+				}
+			} else if strings.HasSuffix(line, variable.SlowLogSQLSuffixStr) {
+				// Get the sql string, and mark the start flag to false.
+				rowMap[variable.SlowLogQuerySQLStr] = types.NewStringDatum(string(hack.Slice(line)))
+				rows = append(rows, rowMap)
+				rowMap = make(map[string]types.Datum, len(slowQueryCols))
+				startFlag = false
+			}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return rows, nil
+}
+
+func parseSlowLogField(field, value string) (*types.Datum, error) {
+	col := findColumnByName(slowQueryCols, field)
+	if col == nil {
+		return nil, errors.Errorf("can't found column %v", field)
+	}
+	var val types.Datum
+	switch col.tp {
+	case mysql.TypeLonglong:
+		num, err := strconv.ParseUint(value, 10, 64)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		val = types.NewUintDatum(num)
+	case mysql.TypeVarchar:
+		val = types.NewStringDatum(value)
+	case mysql.TypeDouble:
+		num, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		val = types.NewDatum(num)
+	case mysql.TypeTiny:
+		// parse bool
+		val = types.NewDatum(value == "true")
+	case mysql.TypeDatetime:
+		t, err := parseTime(value)
+		if err != nil {
+			return nil, errors.Trace(err)
```

Comment:
Error returned by `parseTime()` is generated by `errors.Errorf()`, which automatically adds a stack trace on the error, no need to trace anymore.